orttraining/orttraining/python/training/ortmodule/ort_triton/_codegen.py

```diff
@@ -0,0 +1,476 @@
+            output_var_name = context.get_internal_variable_name(reduce_node.outputs[0].name)
+            tmp_output_var_name = "tmp_" + output_var_name
+            if reduce_node.op_type == "ReduceSum":
+                if not masks_str:
+                    src_code += f"{space_indent}{tmp_output_var_name} = {tmp_output_var_name} + {input_var_name}\n"
+                else:
+                    src_code += (
+                        f"{space_indent}{tmp_output_var_name} = "
+                        f"tl.where({masks_str}, {tmp_output_var_name} + {input_var_name}, {tmp_output_var_name})\n"
+                    )
+            else:
+                op_str = " < " if reduce_node.op_type == "ReduceMax" else " > "
+                if masks_str:
+                    masks_str += " & "
+                src_code += (
+                    f"{space_indent}{tmp_output_var_name} = tl.where("
+                    f"{masks_str}({tmp_output_var_name}{op_str}{input_var_name}), "
+                    f"{input_var_name}, {tmp_output_var_name})\n"
+                )
+        space_indent_outside = " " * (indent - 4)
+        for reduce_node in node.reduce_nodes:
+            output_var_name = context.get_internal_variable_name(reduce_node.outputs[0].name)
+            input_var_name = "tmp_" + output_var_name
+            src_code += (
+                f"{space_indent_outside}{output_var_name} = "
+                f"{reduce_node.triton_func}({input_var_name}, axis=1)[:, None]\n"
+            )
+        return src_code
+
+    def DropoutNode(self, node: DropoutNode, context: CodegenContext, indent: int) -> str:  # noqa: N802
+        space_indent = " " * indent
+        input_var_name = context.get_internal_variable_name(node.inputs[0].name)
+        p_var_name = context.get_internal_variable_name(node.inputs[1].name)
+        output_var_name = context.get_internal_variable_name(node.outputs[0].name)
+        mask_var_name = (
+            context.get_internal_variable_name(node.outputs[1].name)
+            if len(node.outputs) >= 2
+            else "dropout_mask_output"
+        )
+        offset_str = f"{node.global_offset} + " if node.global_offset != sympy.Integer(0) else ""
+        offset_str += self._get_offset_mask(node.offset_calc, node.inputs[0].name)[0]
+        return (
+            f"{space_indent}p = 1 - {p_var_name}\n"
+            f"{space_indent}random = tl.rand(t_seed_cuda, {offset_str})\n"
+            f"{space_indent}{mask_var_name} = random < p\n"
+            f"{space_indent}{output_var_name} = tl.where({mask_var_name}, {input_var_name} / p, 0.0)\n"
+        )
+
+    def ModuleNode(self, node: ModuleNode, context: CodegenContext, indent: int) -> str:  # noqa: N802
+        space_indent = " " * indent
+        code = (
+            f"{space_indent}import triton\n"
+            f"{space_indent}import triton.language as tl\n"
+            f"{space_indent}import torch\n"
+        )
+
+        for kernel_node in node.kernels:
```

Comment:
It is more efficient to use a list and join after or to use a StringIO.

Reply:
This requires big change, I think we can do this refactor later. Since this code runs once during the first step, and the string is not really that long, I think it's OK for now.

Reply:
If you don't want to change your code too much, you can create a class:

```python
class Container:
    def __init__(self):
        self.rows = []
    def __add__(self, s: str):
        self.rows.append(s)
    def __str__(self):
        return "".join(self.rows)
```